man/man1/ccdsky.1
.TH CCDSKY 1NEMO "26 February 2024"

.SH "NAME"
ccdsky \- lazy (cosmology) sky scaling calculator

.SH "SYNOPSIS"
\fBccdsky\fP [parameter=value]

.SH "DESCRIPTION"
\fBccdsky\fP is a lazy sky scaling calculator. It will give you the scaling parameters
for a FITS file (which uses degrees for angles, and m/s for velocities), to make
your typical observer or astronomical software package happy and print out values 
in units observers can be in awe about. It is also useful to remind yourself
how many arcseconds/minutes a given object is at a given distance.
.PP
Although you can transform an image/cube, the default only needs to know the
distance (in AU, pc, kpc, mpc, z, ...) and unit size (in AU, pc, kpc, mpc, ....) of
the object. This will give you the scaling factors for length and velocities.
.PP 
When the distance is given in z, Wright's (2006) cosmology calculator is used.

.SH "PARAMETERS"
.so man1/parameters
.TP 20
\fBin=\fP
Optional input image file. Must be in \fIimage(5NEMO)\fP format.
.TP
\fBout=\fP
Output image file, only required if an input image is given. 
Will be in  \fIimage(5NEMO)\fP format, with optionally scaled image values and
axes coordinates.
.TP
\fBd=#,[unit]\fP
Distance to your object, and optionally a unit.  
AU, pc, Kpc, Mpc, and Gpc are allowed. You can also specify a dimensionless
number 'z', in which case the special cosmology calculator is used. Or if you know
the recession speed and assume a Hubble constant, d=4500/69,Mpc
[default: 1,pc]    
.TP
\fBr=#,[unit]\fP
Length scale of object, and optionally a unit.
km, AU, pc, Kpc, Mpc, and Gpc are allowed. 
Together with a distance \fBd=\fP, this
will convert the length scale in your maps to degrees, required for FITS.
[default: 1,AU]   
.TP
\fBv=#,[unit]\fP
Velocity scale of object, and optionally a unit. The output units will be
the one that FITS uses, m/s.
[Default: 1,km/s]   
.TP
\fBsdv=#\fP
Total flux, in Jy.km/s, of the source. Based on the distance, it will
the compute the HI and H2 (given some reasonable X factor) masses, in
units of solar masses. Default: 1
.TP
\fBfreq=\fP
Observing frequency in GHz [115]
.TP
\fBbeam=\fP
Beam size in arcsec [not used] [12.65]
.TP
\fBdish=\fP
Dish size in meter. For given dish size and frequency, the beam and conversions
from K to Jy/beam is given. [50]
.TP
\fBscale=#\fP
Scale the intensity values in the data itself. Default: 1.
.TP
\fBH=\fP
Specify the cosmology model, with up to 3 parameters. Only used
when units in \fBd=\fP are 'z' were specified.
The first parameter is the Hubble Constant.
Optionally followed by the WM (Omega(matter)) and WV (Omega(vacuum)),
for different cosmologies. If only H0 and WM are given,
a flat universe (WV=0) is used. See also Wright(2006).
WMAP-9 values: H=71,0.27,0.73 (2011?),
Planck-1 values: H=67.15,0.317,0.683 (2013)
[Default:  71,0.27,0.73]
.TP
\fBnsteps=\fP
Max. number of integration steps in the cosmo code to achieve accuracy.
[Default: 1000]


.SH "H"
It will be amusing to see the evolution of H:
.nf
.ta +2i +1i  +1i +1i
WMAP-9 (2011)	71	0.27	0.73
Planck (2013)	67.15	0.317	0.683
Planck (2018)	67.66	0.3111	0.6889
.fi


.SH "EXAMPLES"
Viewing a model at a distance 1 pc, where 1 length unit represents 1 AU,
the coordinates need to be multiplied by 0.000277785 in order for FITS
to view them in the correct angular size of 1 arcsec. 
.nf

% ccdsky d=1,pc r=1,AU v=1,km/s
d=1 pc
r=1 AU
v=1 km/s
rscale=0.000277785  (1.00003 arcsec)
vscale=1000

.fi

To find out the radius of 2 pc at the distance of the galactic center:
.nf
ccdsky r=2,pc d=8.5,kpc
rscale=0.0134814  (48.5329 arcsec)

.fi

Here is an example of creating a small bar, at position angle 30, and
observed at RA=6h and DEC=30d:
.nf

% ccdgen "" map4 bar 1,10,0.5,30 size=512,512,1 
% ccdsky map4 map4b
% ccdfits in=map4b out=map4b.fits radecvel=t crval=90,30 crpix=256.5,256.5
% # now switch to MIRIAD
% fits in=map4b.fits out=map4b.mir op=xyin
% cgdisp in=map4b.mir device=/xs labtyp=arcmin

.fi
and you should see a bar (possibly with a sign error position angle)
of about 1 arcmin in length, in an 8 arcmin field. Notice that
\fIccdfits(1NEMO)\fP also has various options to specify a WCS
which can override the one set by \fIccdgen(1NEMO)\fP.
.PP
Here is a cosmological example:
.nf

% ccdsky H=67.7,0.307,0.693   d=2.19,z
-------------------------------------------------------------
For H_o = 67.7  Omega_M = 0.31 Omega_vac = 0.69 z = 2.190
It is now 13.830 Gyr since the Big Bang.
The age at redshift z was 3.012 Gyr.
The light travel time was 10.818 Gyr.
The comoving radial distance, which goes into Hubbles law, is 5592.2 Mpc or 18.239 Gly
The comoving volume within redshift z is 732.530 Gpc^3.
The angular size distance D_A is 1753.026 Mpc or 5.718 Gly.
This gives a scale of 8.499  kpc/arcsec
The luminosity distance D_L is 17839.0 Mpc or 58.183 Gly.
The distance modulus, m-M, is 46.26
-------------------------------------------------------------
d=2.19 z  [3635.8 Mpc]
r=1 AU
v=1 km/s
SdV=1 Jy.km/s
rscale=1.5846e-13  [ 5.70457e-10 arcsec   5.70457e-07 mas]
vscale=1000
iscale=1
Mass(HI) = 7.25098e+11  
Mass(H2) = 3.22675e+10  (alpha=4.3; includes 1.36 He contribution)

.fi
compare this with the (selected) output for \fBgnuastro\fP's
\fIastcosmiccal(1)\fP program:
.nf
  % astcosmiccal  --H0=67.7 --olambda=0.693 --omatter=0.307 -z2.19
  CosmicCalculator (GNU Astronomy Utilities) 0.11

Universe now
 ------------
 - Age of Universe now (Ga*):                             13.844296
 - Critical density now (g/cm^3):                         8.610662e-30
 - Proper distance to z (Mpc):                            5592.995113
 - Angular diameter distance to z (Mpc):                  1753.290004
 - Tangential distance covered by 1 arcsec at z (Kpc):    8.500190
 - Luminosity distance to z (Mpc):                        17841.654411
 - Distance modulus at z (no unit):                       46.257176
 - Conversion to absolute magnitude (no unit):            44.997699

Universe at desired redshift z
 ------------------------------
 - Age of Universe at z (Ga*):                            3.017860
 - Look-back time to z (Ga*):                             10.826436
 - Critical density at z (g/cm^3):                        9.177897e-29

.fi

.SH "CAVEATS"
Although the conversion from an observed HI flux to mass is not
too controversial (sans optical depth effects), the conversion 
from CO to a mass is still under debate. We use the X value
2e20 cm-2/K km/s.

.SH "SEE ALSO"
ccdgen(1NEMO), ccdmath(1NEMO), ccd2vel(1NEMO), ccdfits(1NEMO), snapscale(1NEMO), image(5NEMO)
.PP
http://arxiv.org/abs/astro-ph/0609593  \fIA Cosmology Calculator for the Web\fP (E.Wright)
.PP
http://www.astro.ucla.edu/~wright/CosmoCalc.html \fI The CosmoCalc website\fP
.PP
https://colossus.astro.umd.edu/ \fIColossus Cosmology Calculator\fP
.PP
https://ui.adsabs.harvard.edu/abs/2013ARA%26A..51..207B/abstract
.PP
https://ascl.net/gnuastro

.SH "FILES"
NEMO/src/image/trans	ccdsky.c - source code

.SH "AUTHOR"
Peter Teuben

.SH "HISTORY"
.nf
.ta +1.5i +5.5i
17-Aug-2012	V1.0 Created		PJT
23-aug-2012	V1.1 added sdv=		PJT
28-aug-2012	V1.2 implemented scale=	PJT
28-feb-2013	V2.2 more verbose, added H=	PJT
16-mar-2013	V3.0 added Wright's cosmology calculator	PJT
21-dec-2022	V3.1 add some Jy/K support		PJT
.fi


src/image/trans/ccdsky.c
/* 
 * CCDSKY: lazy sky scaling calculator to help converting theory data
 *         into degrees and m/s for FITS
 *         Can optionally convert an image. 
 *
 *      17-aug-2012     created        Peter Teuben
 *      22-aug-2012     added sdv=     PJT
 *      28-aug-2012     implemented iscale=
 *      29-jan-2013     allow distance to be in cosmological 'z'
 *      16-mar-2013     3.0 added Wright's CosmoCalculator math, verbose
 *                      (cosmocalc in ASCL?)
 *      18-nov-2022     add Jy/K factor in three different ways
 *
 */


#include <stdinc.h>
#include <getparam.h>
#include <vectmath.h>
#include <filestruct.h>
#include <extstring.h>
#include <image.h>

#include <mks.h>

string defv[] = {
  "in=\n            Optional input image file",
  "out=\n           Output image file",
  "d=1,pc\n         Distance to object, optional unit [pc]",
  "r=1,AU\n         Length scale of object, optional unit [AU]",
  "v=1,km/s\n       Velocity scale of object, optional unit [km/s]",
  "sdv=1\n          Integrated Flux (must be in Jy.km/s)",
  "freq=115\n       Observing frequency in GHz",
  "beam=12.65\n     Beam size in arcsec [not used]",
  "dish=50\n        Dish size in meter",
  "scale=1\n        Scale image values",
  "H=71,0.27,0.73\n Hubble Constant, in case [d] is 'z', with optional WM and WV",
  "nsteps=1000\n    Integrations steps in the cosmo code (accuracy)",
  "VERSION=3.1a\n   22-nov-2022 PJT",
  NULL,
};

string usage = "lazy sky scaling (cosmology) calculator";


#define CVI(x,y,z)  CubeValue(iptr,x,y,z)


#define MAXU 64


/*   
 * See also:  http://www.cv.nrao.edu/course/astr534/HILine.html
 */

static real HI_factor = 2.3595e5;

/* 
 * Xco=2e20 cm-2/K km/s, and alpha_co=4.3 Msun/K km/s 
 * also included 1.36  factor due to Helium contribution to the mass. 
 * Use Xco=14e20 for 13CO, about 7 times that of 12CO
 *
 * Older material is also: http://ned.ipac.caltech.edu/level5/March09/Solomon/Solomon2.html
 */

static real CO_factor = 1.05e4;



/* 
 * CC (cosmology) parameters
 */

static real H0;
static real WM;
static real WV;
static real DA_Mpc;

void setCC(int nc, real *cpar);
void CC(real z, int nsteps, int verbose);



/*
 *  convert a dimensionless 'z' to a distance in Mpc
 *  given a Hubble constant in km/s/Mpc
 */

real z_to_d(real z, real H)
{
  real z1 = (z+1)*(z+1);
  real d = (z1-1)/(z1+1) * c_MKS / H / 1000.0;    /*   d is now Mpc  */
  return d;
}


/*
 *     get_nu :   parse a   NUMBER,UNIT    string
 *     input:   kv         e.g.   "1,pc"
 *              defunit    e.g.   "pc"
 *     output:  value      
 *              unit
 *
 */

void get_nu(string kv, real *value, string unit, string defunit)
{
  string *sp = burststring(kv,",");
  int nsp = xstrlen(sp,sizeof(string)) - 1;
  int nv;

  *value = 1.0;
  strcpy(unit,defunit);
  if (nsp == 0) return;

  nv = nemoinpr(sp[0],value,1);
  if (nv != 1) error("error parsing %s",sp[0]);
  if (nsp == 1) return;

  strcpy(unit,sp[1]);
}

/*
 * efactor:  scaling factor between two common units
 *           u1:    first unit
 *           u2:    second unit
 *           returned value:   how many u1's in a u2  (i.e.   u2/u1)      
 *
 *           @todo need a more general unit + prefix conversion, this
 *                 was just a quick hack
 */

real efactor(string u1, string u2)
{
  real s = -1.0;
  if (streq(u1,"AU")) {
    if (streq(u2,"AU")) 
      s = 1.0;
    else if (streq(u2,"pc")) 
      s = PC/AU;
    else if (streq(u2,"km")) 
      s = 1000.0/AU;
    else if (streq(u2,"Kpc") || streq(u2,"kpc"))
      s = 1e3*PC/AU;
    else if (streq(u2,"Mpc") || streq(u2,"mpc"))
      s = 1e6*PC/AU;
    else
      error("Comparison_1 unit %s not understood for %s",u2,u1);
  } else if (streq(u1,"pc")) {
    if (streq(u2,"pc")) 
      s = 1.0;
    else if (streq(u2,"Kpc") || streq(u2,"kpc"))
      s = 1e3;
    else if (streq(u2,"Mpc") || streq(u2,"mpc"))
      s = 1e6;
    else if (streq(u2,"Gpc") || streq(u2,"Gpc"))
      s = 1e9;
    else if (streq(u2,"AU"))
      s = AU/PC;
    else if (streq(u2,"z")) {
      s = -1.0;  /* to be ammended later */
    } else
      error("Comparison_2 unit %s not understood for %s",u2,u1);
  } else if (streq(u1,"Mpc")) {
    if (streq(u2,"pc")) 
      s = 1e-6;
    else if (streq(u2,"Kpc") || streq(u2,"kpc"))
      s = 1e-3;
    else if (streq(u2,"Mpc") || streq(u2,"mpc"))
      s = 1;
    else if (streq(u2,"Gpc") || streq(u2,"gpc"))
      s = 1e3;
    else if (streq(u2,"z")) {
      s = -1.0;  /* to be ammended later */
    } else
      error("Comparison_3 unit %s not understood for %s",u2,u1);
  } else if (streq(u1,"m/s")) {
    if (streq(u2,"m/s")) 
      s = 1;
    else if  (streq(u2,"km/s")) 
      s = 1e3;
    else
      error("Comparison_4 unit %s not understood for %s",u2,u1);      
  } else
    error("Comparison_5 unit %s not understood",u1);
  dprintf(1,"u1=%s u2=%s  s=%g\n",u1,u2,s);
  return s;
}



void nemo_main(void)
{
    stream  instr, outstr;
    int     ix, iy, iz, nx, ny, nz, nc;
    int     nsteps = getiparam("nsteps");
    imageptr iptr=NULL;
    real d,r,v,  rscale, vscale, iscale = getrparam("scale");
    real mass, sdv_scale,  sdv = getrparam("sdv");
    real cpar[3];
    char ds[MAXU], rs[MAXU], vs[MAXU];

    nc = nemoinpr(getparam("H"), cpar, 3);
    if (nc < 0) error("Parsing cosmo H=%s",getparam("H"));
    setCC(nc, cpar);   /* set H0, WM, WV */

    /*  get the distance and scales      */

    get_nu(getparam("d"),&d,ds,"pc");
    get_nu(getparam("r"),&r,rs,"AU");
    get_nu(getparam("v"),&v,vs,"km/s");

    if (streq(ds,"z")) {
      CC(d,nsteps,0);
      CC(d,nsteps,1);
      printf("d=%g %s  [%g Mpc]\n",d,ds,z_to_d(d,H0));
      // d = z_to_d(d,H0);
      d = DA_Mpc;
      strcpy(ds,"Mpc");
    } else
      printf("d=%g %s\n",d,ds);
    printf("r=%g %s\n",r,rs);
    printf("v=%g %s\n",v,vs);
    printf("SdV=%g Jy.km/s\n",sdv);

    /* angles: convert to deg for FITS */
    rscale = (r * AU) / ( d * PC) * 180.0/PI;   
    rscale *= efactor("AU",rs);
    rscale /= efactor("pc",ds);
    if (rscale < 0) error("z-machine d");
    if (rscale > 0.01666) 
      printf("rscale=%g  [ %g arcsec   %g arcmin]\n",rscale,rscale*3600.0,rscale*60);
    else if (rscale < 2.777e-7)
      printf("rscale=%g  [ %g arcsec   %g mas]\n",rscale,rscale*3600.0,rscale*3600000.0);
    else
      printf("rscale=%g  (%g arcsec)\n",rscale,rscale*3600.0);
    /* velocities:  convert to m/s for FITS */
    vscale = v;
    vscale *= efactor("m/s",vs);
    printf("vscale=%g\n",vscale);

    /* mapvalues */
    printf("iscale=%g\n",iscale);

    /* sdv calculation */
    sdv_scale = efactor("Mpc",ds);
    if (sdv_scale < 0) error("z-machine sdv");
    mass = HI_factor * sqr( d * sdv_scale ) * sdv;
    printf("Mass(HI) = %g  \n",mass);
    mass = CO_factor * sqr( d * sdv_scale ) * sdv;
    printf("Mass(H2) = %g  (alpha=4.3; includes 1.36 He contribution)\n",mass);

    /* jy/k calculations */
    /* See also: https://science.nrao.edu/facilities/vla/proposing/TBconv */
    real freq = getdparam("freq");
    //real beam = getdparam("beam");
    real dish = getdparam("dish");
    real alpha = 1.15;  /* beam = alpha*lambda/D factor */    //  sqrt(3514/3127) = 1.06
    real beam = alpha * 299792.458 * 1e3 / (freq*1e9 * dish) * 206264.81;
    real jyperk  = sqr(freq*beam)/1.222e6;        // NRAO
    real jyper2  = alpha*alpha*3127/(dish*dish);  // mine (a^2*3127/D^2)
    real jyper3  = 8*1380/3.141592/(dish*dish);   // Schloerb using k.T = S.A/2 where A=pi.D^2/4   (3514/D^2)
    printf("Freq %g Dish %g Beam %g alpha %g Jy/K %g %g %g\n",freq,dish,beam,alpha,jyperk,jyper2,jyper3);
      

    if (hasvalue("in") && hasvalue("out")) {      /* patch image if needed */
      instr = stropen(getparam("in"), "r");
      read_image( instr, &iptr);
      Xmin(iptr) *= rscale;   // @todo   review these when Xref != 0
      Ymin(iptr) *= rscale;
      Zmin(iptr) *= vscale;
      Dx(iptr) *= rscale;
      Dy(iptr) *= rscale;
      Dz(iptr) *= vscale;
      Beamx(iptr) *= rscale;
      Beamy(iptr) *= rscale;
      Beamz(iptr) *= vscale;
      // Xref,Yref,Zref no need to change
      if (iscale != 1.0) {
	nx = Nx(iptr);
	ny = Ny(iptr);
	nz = Nz(iptr);
	for (iz=0; iz<nz; iz++)
	  for (iy=0; iy<ny; iy++)
	    for (ix=0; ix<nx; ix++)
	      CubeValue(iptr,ix,iy,iz) *= iscale;
      }

      outstr = stropen(getparam("out"), "w");
      write_image(outstr, iptr);
    }
}



/*
 * taken from Wright's Cosmo Calculator (Schombert's python module)
 * See http://www.astro.ucla.edu/~wright

example: 

CC.py 0.1 71 0.27 0.73
  T      D kpc/arcsec  m-M
  12.38 413.49 1.82 38.29


 */

void setCC(int nc, real *cpar)
{
  if (nc==0) {              // no values, assume Benchmark Model
    H0 = 75.0;
    WM = 0.3;
    WV = 1.0 - WM - 0.4165/(H0*H0);
  } else if (nc == 1) {     // one value, assume Benchmark Model with given Ho
    H0 = cpar[0];
    WM = 0.3;
    WV = 1.0 - WM - 0.4165/(H0*H0);
  } else if (nc == 2) {     // Universe is Open, use Ho, Wm and set Wv to 0.
    H0 = cpar[0];
    WM = cpar[1];
    WV = 0.0;
  } else if (nc == 3) {     // Universe is General, use Ho, Wm and given Wv
    H0 = cpar[0];
    WM = cpar[1];
    WV = cpar[2];
  }
}

void CC(real z, int n, int verbose)
{
  int i;
  real WR, WK, Tyr, DTT, DTT_Gyr, age, age_Gyr, zage, zage_Gyr,
    DCMR, DCMR_Mpc, DCMR_Gyr, DA, DA_Gyr, kpc_DA, DL0, DL_Mpc, DL_Gyr,
    V_Gpc, a, az, h, adot, ratio, x, y, DCMT, VCM, c;
  
  
  /* H0, WM, WV have been set in setCC() */

  /* initialize constants */
  
  WR = 0.0;         // Omega(radiation)
  WK = 0.0;         // Omega curvaturve = 1-Omega(total)
  c = c_MKS/1000.0; // velocity of light in km/sec
  Tyr = 977.8;      // coefficent for converting 1/H into Gyr
  DTT = 0.5;        // time from z to now in units of 1/H0
  DTT_Gyr = 0.0;    // value of DTT in Gyr
  age = 0.5;        // age of Universe in units of 1/H0
  age_Gyr = 0.0;    // value of age in Gyr
  zage = 0.1;       // age of Universe at redshift z in units of 1/H0
  zage_Gyr = 0.0;   // value of zage in Gyr
  DCMR = 0.0;       // comoving radial distance in units of c/H0
  DCMR_Mpc = 0.0;  
  DCMR_Gyr = 0.0; 
  DA = 0.0;         // angular size distance
  DA_Mpc = 0.0;
  DA_Gyr = 0.0;
  kpc_DA = 0.0;
  DL0 = 0.0;        // luminosity distance
  DL_Mpc = 0.0;
  DL_Gyr = 0.0;     // DL0 in units of billions of light years
  V_Gpc = 0.0;
  a = 1.0;          // 1/(1+z), the scale factor of the Universe
  az = 0.5;         //  1/(1+z(object))

  h = H0/100;       // 
  WR = 4.165E-5/(h*h);   //includes 3 massless neutrino species, T0 = 2.72528
  WK = 1-WM-WR-WV;
  az = 1.0/(1+1.0*z);
  age = 0.;
  for (i=0; i<n; i++) {
    a = az*(i+0.5)/n;
    adot = sqrt(WK+(WM/a)+(WR/(a*a))+(WV*a*a));
    age = age + 1./adot;
  }

  zage = az*age/n;
  zage_Gyr = (Tyr/H0)*zage;
  DTT = 0.0;
  DCMR = 0.0;

  /* do integral over a=1/(1+z) from az to 1 in n steps, midpoint rule */

  for (i=0; i<n; i++) {
    a = az+(1-az)*(i+0.5)/n;
    adot = sqrt(WK+(WM/a)+(WR/(a*a))+(WV*a*a));
    DTT = DTT + 1./adot;
    DCMR = DCMR + 1./(a*adot);
  }

  DTT = (1.-az)*DTT/n;
  DCMR = (1.-az)*DCMR/n;
  age = DTT+zage;
  age_Gyr = age*(Tyr/H0);
  DTT_Gyr = (Tyr/H0)*DTT;
  DCMR_Gyr = (Tyr/H0)*DCMR;
  DCMR_Mpc = (c/H0)*DCMR;

  /*  tangential comoving distance */

  ratio = 1.00;
  x = sqrt(fabs(WK))*DCMR;
  if (x > 0.1) {
    if (WK > 0) 
      ratio =  0.5*(exp(x)-exp(-x))/x ;
    else
      ratio = sin(x)/x;
  } else {
    y = x*x;
    if (WK < 0) y = -y;
    ratio = 1. + y/6. + y*y/120.;
  }
  DCMT = ratio*DCMR;
  DA = az*DCMT;
  DA_Mpc = (c/H0)*DA;
  kpc_DA = DA_Mpc/206.264806;
  DA_Gyr = (Tyr/H0)*DA;
  DL0 = DA/(az*az);
  DL_Mpc = (c/H0)*DL0;
  DL_Gyr = (Tyr/H0)*DL0;

  /* comoving volume computation */

  ratio = 1.00;
  x = sqrt(fabs(WK))*DCMR;
  if (x > 0.1) {
    if (WK > 0) 
      ratio = (0.125*(exp(2.*x)-exp(-2.*x))-x/2.)/(x*x*x/3.);
    else
      ratio = (x/2. - sin(2.*x)/4.)/(x*x*x/3.);
  } else {
    y = x*x;
    if (WK < 0) y = -y;
    ratio = 1. + y/5. + (2./105.)*y*y;
  }
  VCM = ratio*DCMR*DCMR*DCMR/3.;
  V_Gpc = FOUR_PI*qbe(0.001*c/H0)*VCM;

  if (verbose) {
    printf("-------------------------------------------------------------\n");
    printf("For H_o = %1.1f  Omega_M = %1.2f Omega_vac = %1.2f z = %1.3f\n",H0,WM,WV,z);
    printf("It is now %1.3f Gyr since the Big Bang.\n", age_Gyr);
    printf("The age at redshift z was %1.3f Gyr.\n",zage_Gyr);
    printf("The light travel time was %1.3f Gyr.\n",DTT_Gyr);
    printf("The comoving radial distance, which goes into Hubbles law, is ");
    printf("%1.1f Mpc or %1.3f Gly\n",DCMR_Mpc ,DCMR_Gyr);
    printf("The comoving volume within redshift z is %1.3f Gpc^3.\n" ,V_Gpc );
    printf("The angular size distance D_A is %1.3f Mpc or %1.3f Gly.\n",DA_Mpc, DA_Gyr);
    printf("This gives a scale of %.3f  kpc/arcsec\n" ,kpc_DA );
    printf("The luminosity distance D_L is %1.1f Mpc or %1.3f Gly.\n",DL_Mpc, DL_Gyr);
    printf("The distance modulus, m-M, is %1.2f\n" , 5*log10(DL_Mpc*1e6)-5 );
    printf("-------------------------------------------------------------\n");
  } else {
    printf("%1.2f " , zage_Gyr);
    printf("%1.2f " , DCMR_Mpc);
    printf("%1.2f " , kpc_DA);
    printf("%1.2f " , (5*log10(DL_Mpc*1e6)-5));
    printf("\n");
  }
}
